File: web/src/api/wire.ts
/**
 * 后端原样发过来的形状。
 *
 * 这些类型必须和 Rust 侧逐字段对齐——它们是照着
 * `cargo run -p lya-core --example wire` 打出的真实 JSON 写的，不是照结构体手抄
 * 的。serde 的重命名、`skip_serializing_if`、枚举 tag 形式手抄很容易错，而这类
 * 偏差要到联调时才炸。改后端消息结构后请重跑那个 example 核对。
 *
 * 这一层**不做任何加工**，加工在 `model/timeline.ts`。分开是因为线上格式由后端
 * 决定，而界面要的形状由界面决定，两者的演进节奏不一样。
 */

/** 消息在树里的角色。 */
export type Role = 'user' | 'assistant' | 'system' | 'tool' | 'hitl'

/** 消息细类。 */
export type MessageKind =
  | 'chat'
  | 'tool_call'
  | 'tool_result'
  | 'form'
  | 'tool_confirm'
  | 'mode_change'
  | 'hitl_response'

/** 消息生命周期。 */
export type MessageStatus =
  | 'streaming'
  | 'pending'
  | 'complete'
  | 'interrupted'
  | 'resolved'

/** 工作模式。 */
export type Mode = 'ask' | 'edit' | 'agent'

/** OpenAI 协议里的一次函数调用。 */
export interface OpenAiToolCall {
  id: string
  type: 'function'
  function: {
    name: string
    /** 未解析的 JSON 字符串——模型可能给出不合法的 JSON，所以保持原样。 */
    arguments: string
  }
}

/** 内嵌的 OpenAI 兼容消息体。HITL 节点没有这一段。 */
export interface OpenAiMessage {
  role: string
  content: string
  tool_calls?: OpenAiToolCall[]
  tool_call_id?: string
}

/** 表单选项。 */
export interface FormOption {
  /** 提交时回传的值。 */
  key: string
  label: string
}

/** 表单里的一道题。 */
export interface FormQuestion {
  id: string
  text: string
  kind: 'single' | 'multi' | 'text'
  /** 文本题为空。 */
  options?: FormOption[]
  /** 是否额外给一个备注输入框。 */
  allow_note?: boolean
}

/** 工具确认里拆出的一步。 */
export interface ConfirmStep {
  raw: string
  explain: string
  /** 没有风险时字段不出现。 */
  risk?: string
  /** 与上一段的关系，如「成功后」。 */
  connector: string
}

/** 需要用户介入的三种情形。 */
export type HitlBlock =
  | { type: 'form'; form_id: string; title: string; questions: FormQuestion[] }
  | {
      type: 'tool_confirm'
      tool_call_id: string
      tool_name: string
      /** 已解析的参数对象——放行后照它执行，所以后端存的是结构化的。 */
      arguments: unknown
      summary: string
      steps?: ConfirmStep[]
      reasons?: string[]
    }
  | { type: 'mode_change'; to_mode: Mode; reason: string }

/** lya 自己的扩展字段。 */
export interface LyaExtras {
  /** 思考全文。落库但不回灌给模型，只用于展示。 */
  reasoning?: string
  /**
   * 后端预留的块列表。
   *
   * **目前它只是 `openai.content` 的镜像**，两边内容完全一样。前端一律以
   * `openai.content` 为准，不读这里——同时读两个会说不清哪个是权威。等后端真的
   * 往里放 content 表达不了的东西（比如富文本卡片）再接。
   */
  blocks?: unknown[]
  hitl?: HitlBlock
  /** 杂项。HITL 解决后用户的原始作答存在 `meta.answer`。 */
  meta?: Record<string, unknown>
}

/** 一条消息的完整载荷。 */
export interface MessagePayload {
  /** schema 版本。 */
  v: number
  role: Role
  kind: MessageKind
  status: MessageStatus
  openai?: OpenAiMessage
  lya: LyaExtras
}

/** 树上的一个节点。 */
export interface MessageRecord {
  id: number
  session_id: string
  /** 根节点为 null。 */
  parent_id: number | null
  /** 会话内递增，给整棵树一个稳定的时间序。 */
  sort_key: number
  payload: MessagePayload
  /** RFC 3339。 */
  created_at: string
}

/** 会话元数据。 */
export interface SessionMeta {
  id: string
  title: string
  status: 'active' | 'archived'
  active_leaf_id: number | null
  work_mode: Mode
  persona: string | null
  /** null 表示用配置里的默认模型。 */
  model_id: string | null
  /** null 表示启用全部工具。 */
  enabled_tools: string[] | null
  created_at: string
  updated_at: string
}

/** 正在跑的那一轮里，一次调用的状态。 */
export interface CallState {
  call_id: string
  name: string
  kind: 'tool' | 'action'
  /** 还在跑时为 null。 */
  ok: boolean | null
}

/**
 * 正在跑的那一轮的实时缓冲。
 *
 * 这些内容还没落库，靠它，刷新页面或换台设备也能接着看当前这轮。
 */
export interface TurnBuffer {
  round: number
  /** 正在写的那条消息节点 id；刚开始时可能还没有。 */
  message_id: number | null
  content: string
  reasoning: string
  calls: CallState[]
}

/** 订阅会话时先收到的快照。 */
export interface Snapshot {
  session: SessionMeta
  /** 当前分支从根到叶，不是整棵树。 */
  messages: MessageRecord[]
  running: TurnBuffer | null
}

/** 整棵树，画分叉图和算分支切换器要用。 */
export interface SessionTree {
  active_leaf_id: number | null
  /** 所有分支端点。 */
  leaves: number[]
  /** 全部节点，按 sort_key 正序。 */
  nodes: MessageRecord[]
}

/** 一轮结束的原因。 */
export type TurnEndReason =
  | { kind: 'completed' }
  | { kind: 'awaiting_human' }
  | { kind: 'max_rounds' }
  | { kind: 'cancelled' }
  | { kind: 'empty_response' }
  | { kind: 'failed'; message: string }

/**
 * SSE 推过来的事件。
 *
 * `message_committed` / `message_updated` / `message_deleted` 三个带着库里发生的
 * 事实，其余是本轮的增量。有了这三个，拿快照起步之后光靠事件流就能维护完整
 * 状态，不必回拉——回拉除了多一次往返，还会让「拉取在路上时新增量到达」变成
 * 一个要处理的竞态。
 */
export type LyaEvent =
  | { type: 'round_started'; round: number }
  | { type: 'reasoning_delta'; text: string }
  | { type: 'message_delta'; text: string }
  | { type: 'message_committed'; record: MessageRecord }
  | { type: 'message_updated'; record: MessageRecord }
  | { type: 'message_deleted'; id: number }
  | { type: 'call_started'; call_id: string; name: string; kind: 'tool' | 'action' }
  | { type: 'call_finished'; call_id: string; name: string; success: boolean }
  | { type: 'await_human'; message_id: number }
  | { type: 'turn_end'; reason: TurnEndReason }

/** 事件信封。 */
export interface Envelope {
  /** `session:<id>` 或 `global`。 */
  scope: string
  type: LyaEvent['type']
  /**
   * 递增序号。
   *
   * 只用于排查问题，**不要依赖它对齐**——快照本身幂等，重连和首次连接走同一
   * 条路，不需要事件重放。
   */
  seq: number
  payload: Record<string, unknown>
}

File: crates/lya-core/examples/wire.rs
//! 把各种消息节点序列化成 JSON 打出来，作为前端类型的依据。
//!
//! 前端类型如果照着 Rust 结构体手抄，serde 的重命名、`skip_serializing_if`、
//! 枚举的 tag 形式都可能对不上，而这类偏差要等联调时才炸。
//!
//! ```bash
//! cargo run -p lya-core --example wire
//! ```

use chrono::{DateTime, Utc};
use lya_agent::{AgentEvent, CallKind, TurnEndReason};
use lya_session::{
    ConfirmStepBlock, FormOption, FormQuestion, FormQuestionKind, HitlBlock, MessageKind,
    MessagePayload, MessageRecord, MessageStatus, OpenAiFunction, OpenAiToolCall,
};

fn dump(label: &str, payload: &MessagePayload) {
    println!("\n// {label}");
    println!("{}", serde_json::to_string_pretty(payload).unwrap());
}

fn main() {
    dump(
        "用户消息",
        &MessagePayload::user_text("帮我看看家目录有多少图片"),
    );

    dump(
        "助手正文（流式中）",
        &MessagePayload::assistant_text("我来看看。", MessageStatus::Streaming),
    );

    let mut with_reasoning = MessagePayload::assistant_text("我来看看。", MessageStatus::Complete);
    with_reasoning.lya.reasoning = Some("用户想知道图片数量，先扫一下目录".into());
    dump("助手正文 + 思考", &with_reasoning);

    // 没有现成的构造器：助手带调用的消息由 agent 从 ChatCompletion 拼出来，
    // 这里照它的形状手搓一个
    let mut calling =
        MessagePayload::assistant_text("先扫一下你的图片目录。", MessageStatus::Complete);
    calling.kind = MessageKind::ToolCall;
    if let Some(openai) = calling.openai.as_mut() {
        openai.tool_calls = Some(vec![OpenAiToolCall {
            id: "call_1".into(),
            kind: "function".into(),
            function: OpenAiFunction {
                name: "image_scan".into(),
                arguments: r#"{"path":"~/图片","recursive":true}"#.into(),
            },
        }]);
    }
    dump("助手发起工具调用（正文与调用可以同时出现）", &calling);

    dump(
        "工具结果（独立节点，靠 tool_call_id 与调用配对）",
        &MessagePayload::tool_result("call_1", "~/图片（共 12 张，列出 12 张）\n…"),
    );

    dump(
        "HITL：表单",
        &MessagePayload::hitl_pending(
            MessageKind::Form,
            HitlBlock::Form {
                form_id: "f_1".into(),
                title: "要删掉这些重复图吗".into(),
                questions: vec![FormQuestion {
                    id: "q1".into(),
                    text: "选择要删除的".into(),
                    kind: FormQuestionKind::Multi,
                    options: vec![FormOption {
                        key: "a".into(),
                        label: "猫副本.png".into(),
                    }],
                    allow_note: true,
                }],
            },
        ),
    );

    dump(
        "HITL：工具确认",
        &MessagePayload::hitl_pending(
            MessageKind::ToolConfirm,
            HitlBlock::ToolConfirm {
                tool_call_id: "call_2".into(),
                tool_name: "bash".into(),
                arguments: serde_json::json!({ "command": "rm ~/图片/猫副本.png" }),
                summary: "执行：rm ~/图片/猫副本.png".into(),
                steps: vec![ConfirmStepBlock {
                    raw: "rm ~/图片/猫副本.png".into(),
                    explain: "删除文件".into(),
                    risk: Some("不可恢复".into()),
                    connector: String::new(),
                }],
                reasons: vec!["会删文件".into()],
            },
        ),
    );

    dump(
        "HITL：模式切换",
        &MessagePayload::hitl_pending(
            MessageKind::ModeChange,
            HitlBlock::ModeChange {
                to_mode: "agent".into(),
                reason: "需要执行命令才能继续".into(),
            },
        ),
    );

    dump(
        "系统标记（用户手动切模式时入树）",
        &MessagePayload::system_text("用户把工作模式切换为 agent。"),
    );

    let mut interrupted =
        MessagePayload::assistant_text("说到一半就被", MessageStatus::Interrupted);
    interrupted.lya.meta = Some(serde_json::json!({ "reason": "cancelled" }));
    dump("被中断的助手消息", &interrupted);

    dump_events();
}

/// SSE 推出去的信封长什么样。
fn dump_events() {
    println!("\n\n// ── SSE 事件 ──────────────────────────────────────");

    let record = MessageRecord {
        id: 7,
        session_id: "s1".into(),
        parent_id: Some(6),
        sort_key: 7,
        payload: MessagePayload::assistant_text("", MessageStatus::Streaming),
        created_at: DateTime::parse_from_rfc3339("2026-08-01T12:00:00Z")
            .unwrap()
            .with_timezone(&Utc),
    };

    let events = [
        AgentEvent::RoundStarted { round: 1 },
        AgentEvent::MessageCommitted {
            record: Box::new(record.clone()),
        },
        AgentEvent::Reasoning("先想想".into()),
        AgentEvent::Delta("好的".into()),
        AgentEvent::CallStarted {
            call_id: "call_1".into(),
            name: "image_scan".into(),
            kind: CallKind::Tool,
        },
        AgentEvent::CallFinished {
            call_id: "call_1".into(),
            name: "image_scan".into(),
            success: true,
        },
        AgentEvent::MessageUpdated {
            record: Box::new(record),
        },
        AgentEvent::MessageDeleted { id: 7 },
        AgentEvent::AwaitHuman { message_id: 8 },
        AgentEvent::TurnEnd {
            reason: TurnEndReason::Failed("HTTP 401".into()),
        },
    ];

    for (seq, event) in events.iter().enumerate() {
        if let Some(envelope) = lya_core::event::from_agent("s1", seq as u64, event) {
            println!("\n// event: {}", envelope.kind);
            println!("{}", serde_json::to_string_pretty(&envelope).unwrap());
        }
    }
}
